mirrord/agent/src/runtime.rs

```diff
@@ -269,13 +270,90 @@ impl ContainerRuntime for ContainerdContainer {
+/// To see agent information, you must:
+/// 1. `kubectl describe {target-pod}`;
+/// 2. `kubectl logs {target-pod} -c {agent-container-name}`;
+#[derive(Debug, Clone)]
+pub(crate) struct EphemeralContainer {
+    /// Used to get the target's `pid`, see [`find_pid_for_ephemeral`] for more information.
+    pub(crate) container_id: String,
+}
+
+/// Searches for the `pid` of the target container when the agent is running in `Mode::Ephemeral`.
+///
+/// It does so by navigating through the dirs in `/proc`, skipping over non-number dirs (like
+/// `/proc/fs`), or when the next dir entry returns an error (since this probably means that
+/// it's not a dir we should be interested in).
+///
+/// To find the `pid`, we look into `/proc/{pid}/cgroup`, searching for a string that matches
+/// `container_id`.
+///
+/// Defaults to returning `1` when we cannot find the `pid`.
+#[tracing::instrument(level = Level::TRACE, ret, err)]
+async fn find_pid_for_ephemeral(container_id: &str) -> Result<u64, tokio::io::Error> {
+    if container_id.is_empty().not() {
+        // The only error we don't ignore here, since if this fails, we can't do the `pid`
+        // search, and this'll break mirrord remote file operations.
+        let mut dir_entries = read_dir("/proc").await?;
+
+        loop {
+            match dir_entries.next_entry().await {
+                Ok(Some(dir)) => {
+                    match dir
+                        .path()
+                        .to_string_lossy()
+                        .split_terminator("/")
+                        .last()
+                        .map(|path| path.parse::<u64>())
+                    {
+                        Some(Ok(potential_pid)) => {
+                            let mut cgroup_path = dir.path();
+                            cgroup_path.push("cgroup");
+
+                            if tokio::fs::read_to_string(cgroup_path)
+                                .await
+                                .map(|read| read.contains(container_id))
+                                .is_ok_and(|found_id| found_id)
+                            {
+                                return Ok(potential_pid);
+                            } else {
+                                tracing::debug!(?potential_pid, "the pid found");
+                                continue;
+                            }
+                        }
+                        // Skip over `parse` errors, since this means that this dir is not a
+                        // `/proc/{pid}`.
+                        _ => {
+                            tracing::debug!("skipping not-number!");
+                            continue;
+                        }
+                    }
+                }
+                Ok(None) => return Ok(1),
+                Err(fail) => {
+                    tracing::warn!(?fail, "Searching for container pid!");
+                    continue;
+                }
+            }
+        }
+    } else {
+        Ok(1)
```

Comment:
We should warn on fallback. (I guess this applies above too)

Reply:
Warnings everywhere now!